### PR TITLE
Parameter-shift vs. backpropagation with default.qubit.tf

### PR DESCRIPTION
**Pull request by josh146**
Adds a tutorial comparing two methods for computing the quantum gradient; the parameter-shift rule (using `default.qubit`) and backpropagation (using `default.qubit.tf`). This tutorial is intended to be more 'practical', highlighting the capabilities of PennyLane when it comes to computing quantum gradients, and showing when different methods should be used.

### CODE REVIEW

**Comment by josh146** [comment]
<h3>Website build</h3>
<strong>Commit:</strong> 8836048053eae3b6b6ee32f89cf6dab03a39c528

<strong>Circle build number:</strong> 851

<strong>Built <code>demos</code> directory:</strong> <a href=https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_demos.zip>https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_demos.zip</a>

<strong>Backreferences zip:</strong> <a href=https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_backreferences.zip>https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_backreferences.zip</a>

<strong>Website zip:</strong> <a href=https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_html.zip>https://851-214003948-gh.circle-artifacts.com/0/tmp/qml_html.zip</a>

<strong>Website url:</strong> <a href=https://851-214003948-gh.circle-artifacts.com/0/_build/html/index.html>https://851-214003948-gh.circle-artifacts.com/0/_build/html/index.html</a>

<em>Please double check the rendered website build to make sure everything is correct.</em>

**Comment by josh146** [comment]
Thanks @mariaschuld! Good points --- I've added in mention of single qubit rotations when introducing the parameter-shift rule, and also swapped out the strongly entangling layers for the basic entangling layers (force of habit 😆 )

**Comment by josh146** [comment]
Hey @mariaschuld, if you don't mind I've switched back to the strongly entangling layers; they have a much high density of parameters per layer, so switching to the basic entangling layers required significantly larger circuit depth to generate the same number of plots. This pushed up the tutorial run time from 4 minutes to 15 minutes

**Comment by josh146** [comment]
@co9olguy

File: demonstrations/tutorial_backprop.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+------------------------
+
+The parameter-shift rules states that, given a variational quantum circuit :math:`U(\boldsymbol
+\theta)` composed of parametrized Pauli rotations, and some measured observable :math:`\hat{B}`, the
+derivative of the expectation value
+
+.. math::
+
+	\langle \hat{B} \rangle (\boldsymbol\theta) =
+	\langle 0 \mid U(\boldsymbol\theta)^\dagger \hat{B} U(\boldsymbol\theta) \mid 0\rangle
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:32]:
```suggestion
	\langle 0 \vert U(\boldsymbol\theta)^\dagger \hat{B} U(\boldsymbol\theta) \vert 0\rangle
```
"`\mid`" causes weird extra spaces: 
image

File: demonstrations/tutorial_backprop.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+@qml.qnode(dev, diff_method="parameter-shift", mutable=False)
+def circuit(params):
+    qml.templates.StronglyEntanglingLayers(params, wires=[0, 1, 2, 3])
+    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2) @ qml.PauliZ(3))
+
+
+##############################################################################
+# Note that we specify that the QNode is **immutable**. This is more restrictive
+# than a standard mutable QNode (the quantum circuit cannot change/differ between
+# executions), however reduces processing overhead.
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:162]:
```suggestion
# executions); however, it reduces processing overhead.

```

File: demonstrations/tutorial_backprop.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+##############################################################################
+# Backpropagation
+# ---------------
+#
+# The parameter-shift rule can be considered a form of `forward-mode autodifferentiation
+# <https://en.wikipedia.org/wiki/forward_accumulation>`__, whereby the function to be differentiated
+# is evaluated multiple times for each independent input variable, gradient rules are applied for
+# each subexpression, and the final gradient accumulated from each of these independent forward
+# sweeps.
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:211]: Is this true? Isn't the standard forward-mode method to use something like dual numbers, where there is only one "forward sweep" (but with twice the size needed for each step in memory)?

Comment (co9olguy) [reply demonstrations/tutorial_backprop.py:211]: (with multiple parameters handled via jacobian-vector product)

File: demonstrations/tutorial_backprop.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+#
+# An alternative to forward-mode autodifferentiation is `reverse-mode autodifferentiation
+# <https://en.wikipedia.org/wiki/Reverse_accumulation>`__; unlike forward-mode differentiation, it
+# requires only a *single* forward pass of the differentiable function to compute the gradient of
+# all variables, at the expense of increased memory usage. During the forward pass, the results and
+# the gradient of all intermediate subexpressions are stored; the computation is then traversed *in
+# reverse*, with the gradient computed by repeatedly applying the chain rule. In most classical
+# machine learning settings, reverse-mode autodifferentiation is the preferred method of
+# autodifferentiation---the reduction in computational time enabling larger and more complex models
+# to be successfully trained. The backpropagation algorithm is a particular special-case of
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:221]: The main advantage of backpropagation, from what I'm aware, is that it is a smarter choice when you have a many-to-few function that you are differentiating (the jacobian/vector can be kept smaller for longer). Forward mode is usually better for differentiating few-to-many functions

File: demonstrations/tutorial_backprop.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+# reverse-mode autodifferentiation, which has helped lead to the machine learning explosion we see
+# today.
+#
+# In quantum machine learning, however, the inability to store and utilize the results of
+# *intermediate* quantum operations on hardware remains a barrier; while reverse-mode
+# autodifferentiation works fine for small quantum simulations, only the forward-mode
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:227]:
```suggestion
# autodifferentiation works fine for small quantum simulations, only the 

```

File: demonstrations/tutorial_backprop.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+# parameter-shift rule can be used to compute gradients on quantum hardware directly. Nevertheless,
+# when training quantum models via classical simulation, it's useful to explore the regimes where
+# reverse-mode differentiation may be a better choice than the parameter-shift rule.
+#
+# Benchmarking
+# ~~~~~~~~~~~~
+#
+# When creating a QNode, :doc:`PennyLane supports various methods of differentiation
+# <code/api/pennylane.qnode>`, including ``"parameter-shift"`` (which we used previously),
+# ``"finite-diff"``, and ``"backprop"``. While ``"parameter-shift"`` works with all devices,
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:237]:
```suggestion
# ``"finite-diff"``, ``"reversible"``, and ``"backprop"``. While ``"parameter-shift"`` works with all devices,

```

File: demonstrations/tutorial_backprop.py [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+# simulator and hardware, ``"backprop"`` will only work for specific simulator devices that are
+# designed to support reverse-mode differentiation.
+#
+# One such device is :class:`default.qubit.tf <pennylane.plugins.default_qubit_tf.DefaultQubitTF>`.
+# This device is a pure state-vector simulator like ``default.qubit``, however unlike
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:242]:
```suggestion
# This device is a pure state-vector simulator like ``default.qubit``; however, unlike

```

File: demonstrations/tutorial_backprop.py [hunk 8]
```diff
@@ -0,0 +1,431 @@
...
+with tf.GradientTape(persistent=True) as tape:
+    res = circuit(params)
+
+times = timeit.repeat("tape.gradient(res, params)", globals=globals(), number=number, repeat=repeat)
+backward_time = min(times) / number
+print(f"Backward pass (best of {repeat}): {backward_time} sec per loop")
+
+##############################################################################
+# Unlike with the parameter-shift rule, the time taken to perform the backwards pass seems
+# independent of the number of differentiable parameters, and instead seems of the order of a single
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:291]: How is this conclusion drawn? The circuit is not compared over different numbers of parameters :thinking:

File: demonstrations/tutorial_backprop.py [hunk 9]
```diff
@@ -0,0 +1,431 @@
...
+ax.legend()
+
+plt.show()
+
+##############################################################################
+# .. raw:: html
+#
+#     <br>
+#
+# We can see that the computational time for the parameter-shift rule increases with
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:392]: In the version I saw, there was some bumpiness in the plot, which might confuse readers. Maybe better to mention something about this, and how it's not really that important to what we're interested in? 

image

File: demonstrations/tutorial_backprop.py [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+    res = circuit(params)
+
+times = timeit.repeat("tape.gradient(res, params)", globals=globals(), number=number, repeat=repeat)
+backward_time = min(times) / number
+print(f"Backward pass (best of {repeat}): {backward_time} sec per loop")
+
+##############################################################################
+# Unlike with the parameter-shift rule, the time taken to perform the backwards pass seems
+# independent of the number of differentiable parameters, and instead seems of the order of a single
+# forward pass!
```
Comment (josh146) [inline demonstrations/tutorial_backprop.py:292]:
```suggestion
# Unlike with the parameter-shift rule, the time taken to perform the backwards pass appears
# of the order of a single forward pass!
```

File: demonstrations/tutorial_backprop.py (hunk 8 above)
Comment (josh146) [reply demonstrations/tutorial_backprop.py:291]: Oh good point. Have removed the incorrect sentence fragment.

File: demonstrations/tutorial_backprop.py (hunk 9 above)
Comment (josh146) [reply demonstrations/tutorial_backprop.py:395]: Have added the following:

```python
# We can see that the computational time for the parameter-shift rule increases with
# increasing number of parameters, as expected, whereas the computational time
# for backpropagation appears much more constant, with perhaps a minute linear increase
# with :math:`p`. Note that the plots are not perfectly linear, with some 'bumpiness' or
# noisiness. This is likely due to low-level operating system jitter, and
# other environmental fluctuations---increasing the number of repeats can help smooth
# out the plot.
#
# For a better comparison, we can scale the time required for computing the quantum
# gradients against the time taken for the corresponding forward pass:
```

File: demonstrations/tutorial_backprop.py (hunk 3 above)
Comment (josh146) [reply demonstrations/tutorial_backprop.py]: @co9olguy have clarified this with a footnote.

File: demonstrations/tutorial_backprop.py (hunk 4 above)
Comment (josh146) [reply demonstrations/tutorial_backprop.py]: Good point! Have modified the language in this paragraph to reflect this.

File: demonstrations/tutorial_backprop.py (hunk 9 above)
Comment (josh146) [reply demonstrations/tutorial_backprop.py:393]: Although the plot is much smoother this build 😆

File: demonstrations/tutorial_backprop.py [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+# ``2*len(params)`` number of times in order to compute the quantum gradient with respect to all
+# parameters. While reasonably fast for a small number of parameters, as the number of parameters in
+# our quantum circuit grows, so does both
+#
+# 1. the circuit depth (and thus the time taken to evaluate each expectation value or 'forward' pass), and
+#
+# 2. the number of parameter-shift evaluations required.
+#
+# Both of these factors increase the time taken to compute the gradient with
+# respect to all parameters (the 'backward' pass).
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:142]:
```suggestion
# respect to all parameters.

```

File: demonstrations/tutorial_backprop.py [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+repeat = 3
+number = 10
+times = timeit.repeat("circuit(params)", globals=globals(), number=number, repeat=repeat)
+forward_time = min(times) / number
+
+print(f"Forward pass (best of {repeat}): {forward_time} sec per loop")
+
+
+##############################################################################
+# We can now time a backwards pass (the time taken to compute all gradients),
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:185]:
```suggestion
# We can now estimate the time taken to compute the full gradient vector,

```

File: demonstrations/tutorial_backprop.py [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+# and see how this compares.
+
+# create the gradient function
+grad_fn = qml.grad(circuit)
+
+times = timeit.repeat("grad_fn(params)", globals=globals(), number=number, repeat=repeat)
+backward_time = min(times) / number
+
+print(f"Backward pass (best of {repeat}): {backward_time} sec per loop")
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:194]:
```suggestion
print(f"Gradient computation (best of {repeat}): {backward_time} sec per loop")

```

File: demonstrations/tutorial_backprop.py [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+
+##############################################################################
+# Backpropagation
+# ---------------
+#
+# The parameter-shift rule can be considered a form of `forward-mode autodifferentiation
+# <https://en.wikipedia.org/wiki/forward_accumulation>`__, whereby the function to be differentiated
+# is evaluated multiple times for each independent input variable, gradient rules are applied for
+# each subexpression, and the final gradient accumulated from each of these independent forward
+# sweeps [#]_.
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:212]:
```suggestion
```

File: demonstrations/tutorial_backprop.py [hunk 15]
```diff
@@ -0,0 +1,438 @@
+r"""
+Quantum gradients with backpropagation
+======================================
+
+.. meta::
+    :property="og:description": Compare and contrast the parameter-shift rule with backpropagation.
```
Comment (co9olguy) [inline demonstrations/tutorial_backprop.py:6]:
```suggestion
    :property="og:description": Using backpropagation can speed up training of quantum circuits compared to the parameter-shift rule—if you are using a simulator.

```
